### PR TITLE
Handle recursive types #324

### LINKED ISSUE #324: Maximum call stack size exceeded
ts-node t.ts
`xxx\node_modules\_typescript@3.7.2@typescript\lib\typescript.js:35954
            function trySymbolTable(symbols, ignoreQualification) {
                                   ^

RangeError: Maximum call stack size exceeded
`

// t.ts
const path = require('path')
const TJS = require('typescript-json-schema')
const glob = require('glob')

// optionally pass argument to schema generator
const settings = {
  required: true,
  uniqueNames: true,
  out: './a.json',
  strictNullChecks: true,
  include: ['**/*schema.ts'],
}

TJS.exec('./tsconfig.json', '*', settings)

// tsconfig.json
`{
  "compilerOptions": {
    /* Basic Options */
    // "incremental": true,                   /* Enable incremental compilation */
    "target": "es5",                          /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017', 'ES2018', 'ES2019' or 'ESNEXT'. */
    "module": "ESNext",                     /* Specify module code generation: 'none', 'commonjs', 'amd', 'system', 'umd', 'es2015', or 'ESNext'. */
    // "lib": [],                             /* Specify library files to be included in the compilation. */
    "allowJs": true,                       /* Allow javascript files to be compiled. */
    // "checkJs": true,                       /* Report errors in .js files. */
    "jsx": "preserve",                     /* Specify JSX code generation: 'preserve', 'react-native', or 'react'. */
    // "declaration": true,                   /* Generates corresponding '.d.ts' file. */
    // "declarationMap": true,                /* Generates a sourcemap for each corresponding '.d.ts' file. */
    "sourceMap": true,                     /* Generates corresponding '.map' file. */
    // "outFile": "./",                       /* Concatenate and emit output to single file. */
    // "outDir": "./",                        /* Redirect output structure to the directory. */
    // "rootDir": "./",                       /* Specify the root directory of input files. Use to control the output directory structure with --outDir. */
    // "composite": true,                     /* Enable project compilation */
    // "tsBuildInfoFile": "./",               /* Specify file to store incremental compilation information */
    // "removeComments": true,                /* Do not emit comments to output. */
    // "noEmit": true,                        /* Do not emit outputs. */
    // "importHelpers": true,                 /* Import emit helpers from 'tslib'. */
    // "downlevelIteration": true,            /* Provide full support for iterables in 'for-of', spread, and destructuring when targeting 'ES5' or 'ES3'. */
    // "isolatedModules": true,               /* Transpile each file as a separate module (similar to 'ts.transpileModule'). */

    /* Strict Type-Checking Options */
    // "strict": true,                           /* Enable all strict type-checking options. */
    // "noImplicitAny": true,                 /* Raise error on expressions and declarations with an implied 'any' type. */
    // "strictNullChecks": true,              /* Enable strict null checks. */
    // "strictFunctionTypes": true,           /* Enable strict checking of function types. */
    // "strictBindCallApply": true,           /* Enable strict 'bind', 'call', and 'apply' methods on functions. */
    // "strictPropertyInitialization": true,  /* Enable strict checking of property initialization in classes. */
    // "noImplicitThis": true,                /* Raise error on 'this' expressions with an implied 'any' type. */
    // "alwaysStrict": true,                  /* Parse in strict mode and emit "use strict" for each source file. */

    /* Additional Checks */
    // "noUnusedLocals": true,                /* Report errors on unused locals. */
    // "noUnusedParameters": true,            /* Report errors on unused parameters. */
    // "noImplicitReturns": true,             /* Report error when not all code paths in function return a value. */
    // "noFallthroughCasesInSwitch": true,    /* Report errors for fallthrough cases in switch statement. */

    /* Module Resolution Options */
    "moduleResolution": "node",            /* Specify module resolution strategy: 'node' (Node.js) or 'classic' (TypeScript pre-1.6). */
    "baseUrl": "./",                       /* Base directory to resolve non-absolute module names. */
    "paths": {
      "@/*": ["src/*"],
      "utils/*": ["src/utils/*"],
      "components/*": ["src/components/*"]
    },                           /* A series of entries which re-map imports to lookup locations relative to the 'baseUrl'. */
    // "rootDirs": [],                        /* List of root folders whose combined content represents the structure of the project at runtime. */
    // "typeRoots": [],                       /* List of folders to include type definitions from. */
    // "types": [],                           /* Type declaration files to be included in compilation. */
    // "allowSyntheticDefaultImports": true,  /* Allow default imports from modules with no default export. This does not affect code emit, just typechecking. */
    "esModuleInterop": true,                   /* Enables emit interoperability between CommonJS and ES Modules via creation of namespace objects for all imports. Implies 'allowSyntheticDefaultImports'. */
    // "preserveSymlinks": true,              /* Do not resolve the real path of symlinks. */
    // "allowUmdGlobalAccess": true,          /* Allow accessing UMD globals from modules. */

    /* Source Map Options */
    // "sourceRoot": "",                      /* Specify the location where debugger should locate TypeScript files instead of source locations. */
    // "mapRoot": "",                         /* Specify the location where debugger should locate map files instead of generated locations. */
    // "inlineSourceMap": true,               /* Emit a single file with source maps instead of having a separate file. */
    // "inlineSources": true,                 /* Emit the source alongside the sourcemaps within a single file; requires '--inlineSourceMap' or '--sourceMap' to be set. */

    /* Experimental Options */
    "experimentalDecorators": true      /* Enables experimental support for ES7 decorators. */
    // "emitDecoratorMetadata": true,         /* Enables experimental support for emitting type metadata for decorators. */
  }
}
`

### PR DESCRIPTION
**Pull request by darkowic**
Please:
- [x] Make your pull request atomic, fixing one issue at a time unless there are many relevant issues that cannot be decoupled.
- [x] Provide a test case & update the documentation in the `Readme.md`

Fixes #324, #339, #246 -> Hopefully ;) Tested on examples from this repository https://github.com/Guidewire/ts-to-json-schema-tests (The ExampleProps).

Feedback appreciated.

### CODE REVIEW

**Comment by darkowic** [comment]
CI fails on TS 3.7:
```
    typeof globalThis
      1) type-globalThis should create correct schema
```
Tips appreciated ;)

**Comment by domoritz** [comment]
Awesome. Thank you so muchy. What's the diff for the globalThis test?

**Comment by darkowic** [comment]
Well, type error...
```
  1 failing
  1) schema
       typeof globalThis
         type-globalThis should create correct schema:
     TypeError: Cannot read property '0' of undefined
      at JsonSchemaGenerator.getClassDefinition (typescript-json-schema.ts:890:63)
      at JsonSchemaGenerator.getTypeDefinition (typescript-json-schema.ts:1247:26)
      at JsonSchemaGenerator.getSchemaForSymbol (typescript-json-schema.ts:1288:26)
      at Object.generateSchema (typescript-json-schema.ts:1515:26)
      at Context.<anonymous> (test/schema.test.ts:27:28)
      at processImmediate (internal/timers.js:443:21)
```
I will try to fix it on Monday...

**Comment by domoritz** [comment]
🤦‍♂️ right, the stacktrace.

**Comment by darkowic** [comment]
Well @domoritz, you solved my problem by removing typescript 3.7 from tests :P Ready to merge 🚀

**Comment by domoritz** [comment]
Thank you

File: test/programs/namespace-deep-1/schema.json [hunk 1]
```diff
@@ -55,26 +56,6 @@
...
-            "$ref": "#/definitions/RootNamespace.SubNamespace.HelperB"
-        }
-    },
-    "required": [
-        "nest",
-        "prev",
-        "propA",
-        "propB"
-    ],
     "type": "object"
```
Comment (domoritz) [inline test/programs/namespace-deep-1/schema.json:59]: Should the type still be here? We already have a $ref property so I think this is wrong.

**Pull request merged**